datasource/ethereum/src/block_stream.rs

```diff
@@ -0,0 +1,875 @@
+                }
+                Some(head_ancestor) => {
+                    // We stopped one block short, so we'll compare the parent hash to the
+                    // subgraph ptr.
+                    if head_ancestor.block.parent_hash == subgraph_ptr.hash {
+                        // The subgraph ptr is an ancestor of the head block.
+                        // We cannot use an RPC call here to find the first interesting block
+                        // due to the race conditions previously mentioned,
+                        // so instead we will advance the subgraph ptr by one block.
+                        // Note that head_ancestor is a child of subgraph_ptr.
+                        Box::new(future::ok(ReconciliationStep::ProcessDescendantBlocks {
+                            from: subgraph_ptr,
+                            descendant_blocks: vec![head_ancestor],
+                        }))
+                    } else {
+                        // The subgraph ptr is not on the main chain.
+                        // We will need to step back (possibly repeatedly) one block at a time
+                        // until we are back on the main chain.
+                        Box::new(future::ok(ReconciliationStep::RevertBlock(subgraph_ptr)))
+                    }
+                }
+            }
+        }
+    }
+
+    fn do_step(
+        &self,
+        step: ReconciliationStep,
+    ) -> Box<Future<Item = ReconciliationStepOutcome, Error = Error> + Send> {
+        let ctx = self.clone();
+
+        // We now know where to take the subgraph ptr.
+        match step {
+            ReconciliationStep::Retry => Box::new(future::ok(ReconciliationStepOutcome::MoreSteps)),
+            ReconciliationStep::Done => Box::new(future::ok(ReconciliationStepOutcome::Done)),
+            ReconciliationStep::RevertBlock(subgraph_ptr) => {
+                // We would like to move to the parent of the current block.
+                // This means we need to revert this block.
+
+                // First, load the block in order to get the parent hash.
+                Box::new(ctx.load_block(subgraph_ptr.hash).and_then(move |block| {
+                    debug!(
+                        ctx.logger,
+                        "Reverting block #{} to get back to main chain",
+                        block.block.number.unwrap();
+                        "block_hash" => format!("{:?}", block.block.hash.unwrap())
+                    );
+
+                    // Produce pointer to parent block (using parent hash).
+                    let parent_ptr = EthereumBlockPointer::to_parent(&block);
+
+                    // Revert entity changes from this block, and update subgraph ptr.
+                    future::result(ctx.subgraph_store
+                        .revert_block_operations(ctx.subgraph_id.clone(), subgraph_ptr, parent_ptr)
+                        .map_err(Error::from)
```

Comment:
Yeah, so this could happen due to database problems. I also wanted to guard against race conditions here, so it could fail if the subgraph ptr is not what was expected (under the current implementation I _think_ this is impossible, assuming only one graph-node runs at a time).

This does fall under the larger block stream retry mechanism however, so an error here will bring us back to `get_next_step`. I think this is the right behaviour since it would only happen if an invariant was broken, in which case we should recheck what the step should be

Reply:
Ok, got it.